test/adapters/gimAdapter.spec.ts

```diff
@@ -0,0 +1,549 @@
+      }
+
+      it("should initialize GeneralIndexModule", async () => {
+        await subject();
+
+        const isInitialized = await setToken.isInitializedModule(setV2Setup.generalIndexModule.address);
+        expect(isInitialized).to.be.true;
+      });
+
+      describe("when the operator is not the caller", async () => {
+        beforeEach(async () => {
+          subjectCaller = approvedCaller;
+        });
+
+        it("should revert", async () => {
+          await expect(subject()).to.be.revertedWith("Must be operator");
+        });
+      });
+    });
+
+    context("when GIM adapter is deployed and system fully set up", async () => {
+      beforeEach(async () => {
+        await gimAdapter.connect(operator.wallet).initialize();
+      });
+
+      describe("#startRebalanceWithUnits", async () => {
+        let subjectComponents: Address[];
+        let subjectTargetUnits: BigNumber[];
+        let subjectPositionMultiplier: BigNumber;
+        let subjectCaller: Account;
+
+        beforeEach(async () => {
+          subjectComponents = [setV2Setup.dai.address, setV2Setup.wbtc.address, setV2Setup.weth.address, setV2Setup.usdc.address];
+          subjectTargetUnits = [ether(50), bitcoin(.005), ether(.15), usdc(50)];
+          subjectPositionMultiplier = ether(.999);
+          subjectCaller = operator;
+        });
+
+        async function subject(): Promise<ContractTransaction> {
+          return await gimAdapter.connect(subjectCaller.wallet).startRebalanceWithUnits(
+            subjectComponents,
+            subjectTargetUnits,
+            subjectPositionMultiplier
+          );
+        }
+
+        it("should correctly set the target units and positionMultiplier", async () => {
+          const [preDaiExecInfo, preWbtcExecInfo, preWethExecInfo, preUsdcExecInfo]: any[] = await getAssetExecutionInfo();
+          const prePositionMultiplier = (await setV2Setup.generalIndexModule.rebalanceInfo(setToken.address)).positionMultiplier;
+
+          expect(preDaiExecInfo.targetUnit).to.eq(ether(100));
+          expect(preWbtcExecInfo.targetUnit).to.eq(bitcoin(.01));
+          expect(preWethExecInfo.targetUnit).to.eq(ether(.1));
+          expect(preUsdcExecInfo.targetUnit).to.eq(ZERO);
+          expect(prePositionMultiplier).to.eq(ether(1));
+
+          await subject();
+
+          const [postDaiExecInfo, postWbtcExecInfo, postWethExecInfo, postUsdcExecInfo] = await getAssetExecutionInfo();
+          const postPositionMultiplier = (await setV2Setup.generalIndexModule.rebalanceInfo(setToken.address)).positionMultiplier;
+
+          expect(postDaiExecInfo.targetUnit).to.eq(subjectTargetUnits[0]);
+          expect(postWbtcExecInfo.targetUnit).to.eq(subjectTargetUnits[1]);
+          expect(postWethExecInfo.targetUnit).to.eq(subjectTargetUnits[2]);
+          expect(postUsdcExecInfo.targetUnit).to.eq(subjectTargetUnits[3]);
+          expect(postPositionMultiplier).to.eq(subjectPositionMultiplier);
+        });
+
```

Comment:
Test case missing where old components are passed out of order?

Reply:
+1